### PR TITLE
app/vmauth: restrict redirects in the OIDC http client to stay within the original request host.

### PR DESCRIPTION
**Pull request by makasim**
Fixes https://github.com/VictoriaMetrics/VictoriaMetrics/security/advisories/GHSA-xxqh-2hcc-9fp6

### CODE REVIEW

File: app/vmauth/oidc.go [hunk 1]
```diff
@@ -136,6 +136,16 @@ type openidConfig struct {
 
 var oidcHTTPClient = &http.Client{
 	Timeout: time.Second * 5,
+	CheckRedirect: func(req *http.Request, via []*http.Request) error {
+		firstHost := via[0].URL.Host
```
[the remark this answers is not in the document]
Comment (makasim) [reply app/vmauth/oidc.go:140]: Keeping redirects within the original host should be enough at this point.

File: app/vmauth/oidc.go [hunk 2]
```diff
@@ -136,6 +136,16 @@ type openidConfig struct {
...
+		if req.URL.Host != firstHost {
```
[the remark this answers is not in the document]
Comment (makasim) [reply app/vmauth/oidc.go]: fixed in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/11616/commits/a32d039aa6344938495af20a14d3ebfa2cf4aa6c

File: app/vmauth/oidc_test.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		if r.URL.Path == "/final" {
+			w.WriteHeader(http.StatusOK)
+			return
+		}
+		http.Redirect(w, r, "/final", http.StatusFound)
+	}, http.StatusOK, "")
+
+	// cross-host redirect must be blocked
+	evilSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		t.Fatalf("request must not reach the evil server")
```
[the remark this answers is not in the document]
Comment (makasim) [reply app/vmauth/oidc_test.go]: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/11616/commits/a32d039aa6344938495af20a14d3ebfa2cf4aa6c

File: app/vmauth/oidc.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (makasim) [reply app/vmauth/oidc.go]: I'd avoid complicating the logic. Checking the host with ports should be fine for now.

File: app/vmauth/oidc.go [hunk 4]
```diff
@@ -136,6 +136,16 @@ type openidConfig struct {
 
 var oidcHTTPClient = &http.Client{
 	Timeout: time.Second * 5,
+	CheckRedirect: func(req *http.Request, via []*http.Request) error {
+		firstHost := via[0].URL.Host
+		if !strings.EqualFold(req.URL.Host, firstHost) {
```
[the remark this answers is not in the document]
Comment (makasim) [reply app/vmauth/oidc.go:141]: That's an edge case, and It's not easy to test. I think it's fine to leave it untested.

**Comment by makasim** [comment]
Tested on my homelab with Google as OIDC server.

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by makasim** [comment]
The bug fix was backported to LTS v1.148 line. It was not backported to LTS v1.136 since OIDC feature was only introduced in v1.138.0
